diff --git a/node_modules/hyperswarm/index.js b/node_modules/hyperswarm/index.js
index cbe2250..e905617 100644
--- a/node_modules/hyperswarm/index.js
+++ b/node_modules/hyperswarm/index.js
@@ -306,7 +306,10 @@ module.exports = class Hyperswarm extends EventEmitter {
       const keepNew = existingIsOutdated || (expectedInitiator === conn.isInitiator)
 
       if (keepNew === false) {
-        existing.sendKeepAlive()
+        if(typeof existing.sendKeepAlive == 'function')
+        {
+          existing.sendKeepAlive()
+        }
         conn.on('error', noop)
         conn.destroy(new Error(ERR_DUPLICATE))
         return
@@ -399,7 +402,10 @@ module.exports = class Hyperswarm extends EventEmitter {
   async _handleNetworkChange () {
     // prioritize figuring out if existing connections are dead
     for (const conn of this._allConnections) {
-      conn.sendKeepAlive()
+      if(typeof conn.sendKeepAlive == 'function')
+      {
+        conn.sendKeepAlive()
+      }
     }
 
     const refreshes = []
